include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -0,0 +1,708 @@
+template <class T>
+std::vector<GivensRotation<T>> applyDeflationToArrays(T rho, T tol, SizeType len, T* d_ptr,
+                                                      const SizeType* i_ptr, T* z_ptr, ColType* c_ptr) {
+  std::vector<GivensRotation<T>> rots;
+  rots.reserve(to_sizet(len));
+
+  SizeType i1 = 0;  // index of 1st element in the Givens rotation
+
+  // Iterate over the indices of the sorted elements in pair (i1, i2) where i1 < i2 for every iteration
+  for (SizeType i2 = 1; i2 < len; ++i2) {
+    // Indices of elements sorted in ascending order
+    SizeType i1s = i_ptr[i1];
+    SizeType i2s = i_ptr[i2];
+
+    T& d1 = d_ptr[i1s];
+    T& d2 = d_ptr[i2s];
+    T& z1 = z_ptr[i1s];
+    T& z2 = z_ptr[i2s];
+    ColType& c1 = c_ptr[i1s];
+    ColType& c2 = c_ptr[i2s];
+
+    // if z1 nearly zero deflate the element and move i1 forward to i2
+    if (std::abs(rho * z1) < tol) {
+      c1 = ColType::Deflated;
+      i1 = i2;
+      continue;
+    }
+
+    // Deflate the second element if z2 nearly zero
+    if (std::abs(rho * z2) < tol) {
+      c2 = ColType::Deflated;
+      continue;
+    }
+
+    // If d1 is not nearly equal to d2, move i1 forward to i2
+    if (!diagonalValuesNearlyEqual(tol, d1, d2, z1, z2)) {
+      i1 = i2;
+      continue;
+    }
+
+    // When d1 is nearly equal to d2 apply Givens rotation
+    T c, s;
+    blas::rotg(&z1, &z2, &c, &s);
+    d1 = d1 * c * c + d2 * s * s;
+    d2 = d1 * s * s + d2 * c * c;
+
+    rots.push_back(GivensRotation<T>{i1s, i2s, c, s});
+    //  Set the the `i1` column as "Dense" if the `i2` column has opposite non-zero structure (i.e if
+    //  one comes from Q1 and the other from Q2 or vice-versa)
+    if ((c1 == ColType::UpperHalf && c2 == ColType::LowerHalf) ||
+        (c1 == ColType::LowerHalf && c2 == ColType::UpperHalf)) {
+      c1 = ColType::Dense;
+    }
+    c2 = ColType::Deflated;
+  }
+
+  return rots;
+}
+
+template <class T>
+std::vector<GivensRotation<T>> applyDeflation(
+    SizeType n, pika::shared_future<T> rho_fut, pika::shared_future<T> tol_fut,
+    std::vector<pika::future<matrix::Tile<T, Device::CPU>>> d_tiles,
+    std::vector<pika::shared_future<matrix::Tile<const SizeType, Device::CPU>>> perm_d_tiles,
+    std::vector<pika::future<matrix::Tile<T, Device::CPU>>> z_tiles,
+    std::vector<pika::future<matrix::Tile<ColType, Device::CPU>>> coltypes_tiles) {
+  TileElementIndex zero_idx(0, 0);
+  T* d_ptr = d_tiles[0].get().ptr(zero_idx);
+  T* z_ptr = z_tiles[0].get().ptr(zero_idx);
+  ColType* c_ptr = coltypes_tiles[0].get().ptr(zero_idx);
+  const SizeType* i_ptr = perm_d_tiles[0].get().ptr(zero_idx);
+
+  return applyDeflationToArrays(rho_fut.get(), tol_fut.get(), n, d_ptr, i_ptr, z_ptr, c_ptr);
+}
+
+// Partitions `p_ptr` based on a `ctype` in `c_ptr` array.
+//
+// Returns the number of elements in the second partition.
+inline SizeType partitionColType(SizeType len, ColType ctype, const ColType* c_ptr, SizeType* i_ptr) {
```

Comment:
What is a typical value of `len` here?

Reply:
This can be as large as the number of rows or columns of a distributed matrix. That is, for a big run of a dense 10^6 x 10^6 matrix, len = 10^6.